### DISCUSSION TITLE
Redesign Development Environment

### DISCUSSION THREAD

**Discussion by TheRealGramdalf** [UnicornsOnLSD/finamp#582] [General]
_Migrated from https://github.com/jmshrv/finamp/issues/220#issuecomment-1908904531_

With the redesign beta coming soon™, I'm eager to get started early - and as such, I need to set up a development environment so I can build an APK to test things out. As it currently stands, I was unable to find much information in the contributing guidelines - which essentially say "Install flutter".

Since it seems like StayBlue has already got an environment working, as has Chaphasilor, I'd greatly appreciate some general directions so I can make it work with my current workstation running NixOS. As I currently understand it, the steps required are:

- Install flutter
- Clone the git repository from the redesign branch (`git clone https://github.com/jmshrv/finamp -b redesign`)
- ? Install the android SDK
- ? Run `flutter build apk` to create an installable APK to use for testing and debugging

I've tried to do this already, but NixOS doesn't play nicely with dynamically linked executables, so it's a little hard to make sure I have all the packages needed - any more precise input (along with versions, exact package names and repos if possible) would point me in the right direction.

To the second point: Once I've got this set up, it will be possible to maintain a Nix configuration file that essentially defines all the steps required, such that one could create a development environment with the correct dependencies quickly and easily. For a better explanation, see the beginning of this [blog post](https://ayats.org/blog/nix-workflow/) 
Keep in mind that I'm not super familiar with how flutter works, and whether the idea of consistent build environment really affects this type of project - but it's simple enough to implement (once I've got my environment set up) that it could make contributing much easier for other people to dip their toes in. If that sounds interesting, let me know and we'll work it out from there.

_Original text for reference:_
> I'm hoping to get a dev environment running so I can build the redesign branch (as I mentioned in #354), and I was wondering if you wouldn't mind giving me a hand - I took a look at the Contributing guide, but because I run NixOS things aren't quite as simple. 
StayBlue - you mentioned that you had an environment running that you used to build for iOS in the new downloader PR - if you wouldn't mind sharing the general steps I'd be very appreciative - I can work on porting that to a nixified configuration that works on my machine.
Chaphasilor - If you're interested, I'll see what I can do to make said nixified configuration usable in general - if you haven't heard of NixOS or Nix before, it allows you to have a repeatable environment (package versions, environment variables, etc) such that you can track the nix configuration directly in the git repository, run `nix develop` on any machine with nix installed, and then have the exact same version of flutter, android sdk, or otherwise, regardless of whatever distribution you use. You can also use it to actually build the release APK, which could help people contribute and test more easily.
>But I digress- that's a gross oversimplification, but I believe it gets the point across. You might also think of it as similar to docker images, but much more advanced and precise.
>Cheers!
>~Gramdalf

**Comment by StayBlue** [comment]
Hello! Apologies for the late response. I haven't managed to set up an Android build environment yet. I've only set one for iOS using XCode and the Flutter SDK. However, once I get around to setting up the Android build environment, I might comment on how I set it up on my Mac machine if anyone else would find it beneficial. I'm sorry that I wasn't able to help.

**Reply by UnicornsOnLSD** [reply]
That'd be great :)

**Comment by UnicornsOnLSD** [comment]
Hi, thanks for being eager and willing to test :)

I didn't really cover installing Flutter because I didn't want to handle keeping it up to date/accurate as Flutter evolves. Getting the Android SDK playing nicely is honestly pretty messy even with the official guides, so I doubt I'd be able to write anything useful.

That said, [it looks like Nix has good support for Flutter](https://nixos.wiki/wiki/Flutter), including setting up Android with it. I've never used Nix myself (although the concept sounds cool), so it'd be interesting to hear how you get on there.

If you can get a Nix config working, it'd be great to have so that other people can easily get started with development.

**Reply by Chaphasilor** [reply]
That also looks pretty up-to-date :)

**Reply by TheRealGramdalf** [reply]
Finally found some time to work on this, here's what I've got so far:

```nix
{
description = "Flutter 3.16.x: Finamp devel environment ";
inputs = {
  nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
  flake-utils.url = "github:numtide/flake-utils";
};
outputs = { self, nixpkgs, flake-utils }:
  flake-utils.lib.eachDefaultSystem (system:
    let
      pkgs = import nixpkgs {
        inherit system;
        config = {
          android_sdk.accept_license = true;
          allowUnfree = true;
        };
      };
      buildToolsVersion = "34.0.0";
      androidComposition = pkgs.androidenv.composeAndroidPackages {
        buildToolsVersions = [ "30.0.3" ];
        platformVersions = [ "31" "33" "30" ]; # If flutter build throws errors about platforms;android-33 Android SDK Platform 33, add them here
        abiVersions = [ "armeabi-v7a" "arm64-v8a" ];
      };
      androidSdk = androidComposition.androidsdk;
    in
    {
      devShell =
        with pkgs; mkShell rec {
          GRADLE_OPTS = "-Dorg.gradle.project.android.aapt2FromMavenOverride=${androidSdk}/libexec/android-sdk/build-tools/30.0.3/aapt2";
          ANDROID_SDK_ROOT = "${androidSdk}/libexec/android-sdk";
          buildInputs = [
            flutter
            androidSdk
            jdk17
          ];
        };
    });
}
```

It looks like at this point all I need is some missing environment variables/android SDK plugins - if anyone who has a working (android) build setup can give me a list of plugins and such that would be great. 

In any case, my current error is about `aapt2` - it looks like I might have an old version or something, since it complains about an option that doesn't exist:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':android_id:verifyReleaseResources'.
> Could not resolve all files for configuration ':android_id:releaseRuntimeClasspath'.
   > Failed to transform core-1.6.0.aar (androidx.core:core:1.6.0) to match attributes {artifactType=android-compiled-dependencies-resources, org.gradle.category=library, org.gradle.dependency.bundling=external, org.gradle.libraryelements=aar, org.gradle.status=release, org.gradle.usage=java-runtime}.
      > Execution failed for AarResourcesCompilerTransform: /home/gramdalf/.gradle/caches/transforms-3/d607ab372f7bf69bc81c893ec6808ba5/transformed/core-1.6.0.
         > Android resource compilation failed
           ERROR:AAPT: unknown option '--source-path'
```

For reference, the nixpkgs manual has a good section about how to add plugins then nix way:
https://nixos.org/manual/nixpkgs/unstable/#android

**Reply by TheRealGramdalf** [reply]
Changing the gradle option environment variable to `34.0.0` instead of `30.0.3` worked. Now I've got this:

```nix
{
description = "Flutter 3.16.x: Finamp devel environment ";
inputs = {
  nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
  flake-utils.url = "github:numtide/flake-utils";
};
outputs = { self, nixpkgs, flake-utils }:
  flake-utils.lib.eachDefaultSystem (system:
    let
      pkgs = import nixpkgs {
        inherit system;
        config = {
          android_sdk.accept_license = true;
          allowUnfree = true;
        };
      };
      buildToolsVersion = "34.0.0";
      androidComposition = pkgs.androidenv.composeAndroidPackages {
        buildToolsVersions = [ "30.0.3" "34.0.0" ];
        platformVersions = [ "31" "33" "30" ]; # If flutter build throws errors about platforms;android-33 Android SDK Platform 33, add them here
        abiVersions = [ "armeabi-v7a" "arm64-v8a" ];
      };
      androidSdk = androidComposition.androidsdk;
    in
    {
      devShell =
        with pkgs; mkShell rec {
          GRADLE_OPTS = "-Dorg.gradle.project.android.aapt2FromMavenOverride=${androidSdk}/libexec/android-sdk/build-tools/34.0.0/aapt2";
          ANDROID_SDK_ROOT = "${androidSdk}/libexec/android-sdk";
          buildInputs = [
            flutter
            androidSdk
            jdk19
          ];
        };
        shellHook = ''
          flutter pub get
        '';
    });
}
```
That works for `flutter build apk --debug`, but fails with `flutter build apk` - `SigningConfig "release" is missing required property "storeFile"`.
I assume this is because it needs a key to sign the released APK, so I'm not sure what to do for a development environment.

**Reply by Chaphasilor** [reply]
Awesome! And this is where the contribution guidelines come into play: https://github.com/jmshrv/finamp/blob/main/CONTRIBUTING.md#android-keys 😁
